enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,322 @@
+HAProxy tries to compress all responses that are not compressed by the server.  Therefore, at least one MIME type is
+required in the `HTTPCompressionPolicy`.  We consider the fact that cluster admins may want to create and use their
+own MIME type, but the allowed set of characters is defined in a kubebuilder validation
+pattern  of `"[a-zA-Z0-9\\+-;= ]+/([a-zA-Z0-9\\\+-;= ]+|\\*)"`, to allow a token of alphanumeric plus "_-;= " followed
+by "/" followed by another alphanumeric token of the same pattern, or "*", the wildcard character.
+
+### Open Questions
+Is the MIME type pattern still too strict?  I reviewed the `mime` golang package at https://golang.org/src/mime/type.go,
+`builtinTypesLower` map, but if we intend to allow any custom MIME type, perhaps a wider net should be cast, as in
+https://golang.org/src/mime/grammar.go.
+
+### Test Plan
+- Unit tests will validate
+  - that at least one HTTP compression MIME type exists and any listed MIME types adhere to the
+  correct pattern
+  - that the HTTP compression directive appears in the `defaults` section of the HAProxy template
+  when it is enabled and properly configured on the IngressController spec
+  - that a downgrade to version 4.9 (or version previous to feature implementation version) will not retain the
+  compression configuration of 4.10 (the environment variables should no longer be exposed)
+- E2e tests will validate
+  - that the IngressController spec accepts and enables compression in the `defaults` section of
+  the generated haproxy.cfg
+  - that only requested MIME types receive compression
+
+Note that HAProxy Enterprise **disables** HTTP compression when:
+- The request does not advertise a supported compression algorithm in the "Accept-Encoding" header
+- The response message is not HTTP/1.1 or above
+- The HTTP status code is not one of 200, 201, 202 or 203
+- The response contains neither a "Content-Length" header, nor a "Transfer-Encoding" header whose last value is `chunked`
+- The response contains a "Content-Type" header whose first value starts with `multipart`
+- The response contains a "Cache-Control" header with a value of `no-transform`
+- The response contains a "Content-Encoding" header, indicating that the response is already compressed (see `compression offload`)
+- The response contains an invalid "ETag" header or multiple "Etag" headers
+
+### Graduation Criteria
+
+This feature will be immediately available in release 4.10, and will not have a dev or tech preview.
+
+#### Dev Preview -> Tech Preview
+n/a
+
+#### Tech Preview -> GA
+n/a
+
+#### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+An upgrade from a previous installation that had the compression directives exposed, will need to change to use the
+IngressController spec.
```

Comment:
Previous versions did not expose the compression directives.  The API is additive and turned off by default, so there should be no change in behavior or need for migration during upgrades.

Reply:
Fixed.